Test ipam on secondary networks

alonSadan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

It should be verified that CNI IPAM functions well on secondary networks with bridge binding.
This PR adds a case to the existing bridge binding test to verify that.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

alonSadan commented:
The tests I added [fail on the 1.20 network lane](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/5826/pull-kubevirt-e2e-k8s-1.20-sig-network/1404521972959809536) I think it is related to https://github.com/kubevirt/kubevirt/issues/5477  since the ips are configured as secondary they way it is implemented now.
I converted the PR to draft for now and re-implement.

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-network

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -331,26 +331,41 @@ var _ = SIGDescribe("[Serial]Multus", func() {
+				withVmiOptions := []libvmi.Option{
+					libvmi.WithCloudInitNoCloudNetworkData(cloudInitNetworkData, false),
+				}
+				for idx, _ := range networks {
```
I'm quite sure you can drop the `, _` part here; you can use just an index when iterating a slice.

alonSadan replied:
Fixed

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-network

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-network

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-network

alonSadan commented:
> The tests I added [fail on the 1.20 network lane](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/5826/pull-kubevirt-e2e-k8s-1.20-sig-network/1404521972959809536) maybe it is related to #5477 since the ips are configured as secondary they way it is implemented now.
> I converted the PR to draft for now and re-implement.

The problem was that I (stupidly) didn't make sure the VMIs are on the same node. Now it passes CI and ready for review.
@maiqueb  would you take a look?

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -332,26 +332,41 @@ var _ = SIGDescribe("[Serial]Multus", func() {
 		})
 
 		Context("VirtualMachineInstance with Linux bridge plugin interface", func() {
+			getFedoraVmi := func(interfaces []v1.Interface, networks []v1.Network, cloudInitNetworkData string) *v1.VirtualMachineInstance {
+				withVmiOptions := []libvmi.Option{
+					libvmi.WithCloudInitNoCloudNetworkData(cloudInitNetworkData, false),
+				}
+				for idx := range networks {
+					withVmiOptions = append(withVmiOptions,
+						libvmi.WithInterface(interfaces[idx]),
+						libvmi.WithNetwork(&networks[idx]),
+						libvmi.WithCloudInitNoCloudUserData(tests.GetFedoraToolsGuestAgentUserData(), false),
+					)
+				}
+				return libvmi.NewTestToolingFedora(withVmiOptions...)
+			}
```
I don't quite understand why you can't achieve this with an alpine VM. You'll have to provide a reason for changing the VM image.

alonSadan commented:
> I don't quite understand why you can't achieve this with an alpine VM. You'll have to provide a reason for changing the VM image.

I moved back to use alpine image.  For that to work with IPAM and DHCP I had to console the VMI to setup the interfaces with [setup-interfaces](https://wiki.alpinelinux.org/wiki/Alpine_setup_scripts#setup-interfaces). Otherwise the interfaces won't get configured. 
I think this actually could be better to configure the interfaces with alpine setup script even with static address, as this is the [official way](https://wiki.alpinelinux.org/wiki/Configure_Networking) to configure interfaces in alpine.

anyway, the changes are [here](https://github.com/kubevirt/kubevirt/compare/4e4349652dbb2d7c82775e5bdfb0d3092a154fed..0383798120d1346dff8fa7cd87121b3621c28f82)

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -332,26 +353,43 @@ var _ = SIGDescribe("[Serial]Multus", func() {
 
 		Context("VirtualMachineInstance with Linux bridge plugin interface", func() {
-			table.DescribeTable("should be able to ping between two vms", func(interfaces []v1.Interface, networks []v1.Network, ifaceName string) {
+			table.DescribeTable("should be able to ping between two vms", func(interfaces []v1.Interface, networks []v1.Network, ifaceName, cidr1, cidr2 string) {
```
I wouldn't use `cidr1` and `cidr2` as the variable names here - I would use `staticIPVm1` and `staticIPVm2`. You could pass an empty string when you want to have dhcp (or even the dhcp string you're using now).

alonSadan replied:
Done.

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -332,26 +353,43 @@ var _ = SIGDescribe("[Serial]Multus", func() {
 		Context("VirtualMachineInstance with Linux bridge plugin interface", func() {
-			table.DescribeTable("should be able to ping between two vms", func(interfaces []v1.Interface, networks []v1.Network, ifaceName string) {
+			table.DescribeTable("should be able to ping between two vms", func(interfaces []v1.Interface, networks []v1.Network, ifaceName, cidr1, cidr2 string) {
+				if cidr2 == "dhcp" {
```
What if it's `cidr1` that holds dhcp ? No need to provision the NAD ?

alonSadan replied:
oops. Thanks for the catch.
changed to 
`staticIPVm1 == "" | | staticIPVm2  == "`

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -332,26 +353,43 @@ var _ = SIGDescribe("[Serial]Multus", func() {
+				if cidr2 == "dhcp" {
+					By("ping between virtual machines")
+					// with DHCP we don't know which VMI got which address from the range, so we ping both
+					Expect(libnet.PingFromVMConsole(vmiOne, "10.1.1.2")).To(Succeed())
+					Expect(libnet.PingFromVMConsole(vmiOne, "10.1.1.3")).To(Succeed())
```
Where did these come from ? 

You'll need to come up with a function to retrieve an IP address from the range assigned by IPAM: https://github.com/kubevirt/kubevirt/pull/5826/commits/0383798120d1346dff8fa7cd87121b3621c28f82#diff-da14ca97e77db6bb275427ac69187a539c58fc297e64519628e5385f9f411c5fR358

alonSadan replied:
I decided to remove the range and retrieve the address from the pod with this [function](https://github.com/kubevirt/kubevirt/compare/716efb2f60a6e11b8da900adcc6c3ea080a58e77..3e98959bc57018af7cb77a6cf59adb0b399b3c1e#diff-da14ca97e77db6bb275427ac69187a539c58fc297e64519628e5385f9f411c5fR356-R375)

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -94,6 +94,13 @@ var _ = SIGDescribe("[Serial]Multus", func() {
 		},
 	}
 
+	linuxBridgeInterfaceWithIPAM := v1.Interface{
+		Name: "linux-bridge-with-ipam",
+		InterfaceBindingMethod: v1.InterfaceBindingMethod{
+			Bridge: &v1.InterfaceBridge{},
+		},
```
I see nothing in the description to justify the IPAM name.

alonSadan replied:
In the future it will be associated with a network that has IPAM.  I can move it closer to the place where this association happens (in the test it self), so that it will be clearer. I can also leave it here and rename it to "linux-bridge-secondary" or something like that, it will be more correct in this context, but less descriptive I think.
Which one do you think is better?

maiqueb replied:
It's fine, you can keep it.

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -110,6 +117,25 @@ var _ = SIGDescribe("[Serial]Multus", func() {
 		},
 	}
 
+	linuxBridgeWithIPAMNetwork := v1.Network{
+		Name: "linux-bridge-with-ipam",
+		NetworkSource: v1.NetworkSource{
+			Multus: &v1.MultusNetwork{
+				NetworkName: "linux-bridge-net-ipam",
+			},
```
I see nothing in the description to justify the IPAM name.

alonSadan replied:
In the future it will be associated with a network that has IPAM. I can move it closer to the place where this association happens (in the test it self), so that it will be clearer. I can also leave it here and rename it to "linux-bridge-secondary" or something like that, it will be more correct in this context, but less descriptive I think.
Which one do you think is better?

maiqueb replied:
It's fine.

maiqueb reviewed (changes requested):
Looks good, minor changes requested.

alonSadan commented:
Addressed Miguel [comments](https://github.com/kubevirt/kubevirt/compare/716efb2f60a6e11b8da900adcc6c3ea080a58e77..3e98959bc57018af7cb77a6cf59adb0b399b3c1e). 
Then I forgot to go fmt the file so I [pushed again](https://github.com/kubevirt/kubevirt/compare/3e98959bc57018af7cb77a6cf59adb0b399b3c1e..9c8985490e52f7986d24c606ff41d0165763ad83)

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -332,26 +353,61 @@ var _ = SIGDescribe("[Serial]Multus", func() {
+				}
+
+				return cidrToIP(r.FindString(output))
+			}
```
Can't say I like this part (retrieving the IP via console), but unfortunately, it's either that, or rely on the guest agent (which you'd have to install).

This would never the less be a reason for you to use a fedora based image (we have an img that comes w/ guest agent pre-installed). Should have thought of that before.

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -332,26 +353,61 @@ var _ = SIGDescribe("[Serial]Multus", func() {
+				if err != nil {
+					return "", err
+				}
+				r, err := regexp.Compile(`[0-9]+\.[0-9]+\.[0-9]+\.[0-9]/[0-9]+`)
```
I think this regex is too lenient.

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -332,26 +353,61 @@ var _ = SIGDescribe("[Serial]Multus", func() {
+
+			table.DescribeTable("should be able to ping between two vms", func(interfaces []v1.Interface, networks []v1.Network, ifaceName, staticIPVm1, staticIPVm2 string) {
+				if staticIPVm2 == "" || staticIPVm1 == "" {
+					ipam := "\\\"type\\\": \\\"host-local\\\", \\\"subnet\\\": \\\"10.1.1.0/24\\\""
```
How about you encapsulate this in a short function ? Something like 
```golang
func generateIPAMConfig(type string, subnet string) string {...}
```

maiqueb reviewed (changes requested):
I start to have doubts if using a fedora img wouldn't be preferable, since it allows you to extract the IP address from the vmi status. I have personally chosen that route in the past when given the choice. Sorry for failing to notice that in the previous review.

I would check out what @AlonaKaplan 's thoughts on this.

AlonaKaplan commented:
> I start to have doubts if using a fedora img wouldn't be preferable, since it allows you to extract the IP address from the vmi status. I have personally chosen that route in the past when given the choice. Sorry for failing to notice that in the previous review.
> 
> I would check out what @AlonaKaplan 's thoughts on this.

I think you discovered a bug.
In the described case, reporting the ip in the status should work even if there is no guest agent, we know the ip, we even store it in the cache.
I thing the but is [here](https://github.com/kubevirt/kubevirt/blob/9f046bb678739f1e727d746eeee9d3040f514b79/pkg/virt-handler/vm.go#L877). Try to remove the check checking i it is a pod network.

maiqueb commented:
> > I start to have doubts if using a fedora img wouldn't be preferable, since it allows you to extract the IP address from the vmi status. I have personally chosen that route in the past when given the choice. Sorry for failing to notice that in the previous review.
> > I would check out what @AlonaKaplan 's thoughts on this.
> 
> I think you discovered a bug.
> In the described case, reporting the ip in the status should work even if there is no guest agent, we know the ip, we even store it in the cache.
> I think the bug is [here](https://github.com/kubevirt/kubevirt/blob/9f046bb678739f1e727d746eeee9d3040f514b79/pkg/virt-handler/vm.go#L877). Try to remove the check checking i it is a pod network.

I honestly thought that was intentional  `¯\_(ツ)_/¯` . 

+1 on checking if removing the pod network check gives you the IP address on the VMI status. 

If not, would you mind opening an issue @alonSadan ?

alonSadan commented:
> > > I start to have doubts if using a fedora img wouldn't be preferable, since it allows you to extract the IP address from the vmi status. I have personally chosen that route in the past when given the choice. Sorry for failing to notice that in the previous review.
> > > I would check out what @AlonaKaplan 's thoughts on this.
> > 
> > 
> > I think you discovered a bug.
> > In the described case, reporting the ip in the status should work even if there is no guest agent, we know the ip, we even store it in the cache.
> > I think the bug is [here](https://github.com/kubevirt/kubevirt/blob/9f046bb678739f1e727d746eeee9d3040f514b79/pkg/virt-handler/vm.go#L877). Try to remove the check checking i it is a pod network.

Thank you @AlonaKaplan !
[I removed the limitation](https://github.com/kubevirt/kubevirt/compare/9c8985490e52f7986d24c606ff41d0165763ad83..cf15709306990537bff5c0e7c17ddded14332d16#diff-7a3f9e0461f554b07ac195375eb0f8db7d46579071795a65c459c721b1a8204aR877-R886) , It worked and the Ips are now reported in the status.



> 
> I honestly thought that was intentional `¯\_(ツ)_/¯` .
Yes, me too :)
> 
> +1 on checking if removing the pod network check gives you the IP address on the VMI status.
> 
> If not, would you mind opening an issue @alonSadan ?

Thanks @maiqueb 
The IPs are now reported in the status, [and retrieved from there.](https://github.com/kubevirt/kubevirt/compare/9c8985490e52f7986d24c606ff41d0165763ad83..cf15709306990537bff5c0e7c17ddded14332d16#diff-da14ca97e77db6bb275427ac69187a539c58fc297e64519628e5385f9f411c5fR356-R365)

alonSadan commented:
[fixed a typo](https://github.com/kubevirt/kubevirt/compare/cf15709306990537bff5c0e7c17ddded14332d16..01525c33bcd672188d25ced737770b145a994023)

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -28,6 +28,8 @@ import (
 	"strings"
 	"time"
 
+	"github.com/onsi/ginkgo/extensions/table"
```
nit: shouldn't this be in the block below, after `. "github.com/onsi/ginkgo"` ?

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -121,15 +147,10 @@ var _ = SIGDescribe("[Serial]Multus", func() {
-			Body([]byte(fmt.Sprintf(linuxBridgeConfCRD, "linux-bridge-net-vlan100", tests.NamespaceTestDefault))).
-			Do(context.Background())
-		Expect(result.Error()).NotTo(HaveOccurred())
+		Expect(createBridgeNetworkAttachementDefinition("linux-bridge-net-vlan100", "100", ""))
```
Ok, you `Expect(...)` this to what ? You're missing the matcher part - i.e.`To(Succeed())` .

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -332,26 +353,55 @@ var _ = SIGDescribe("[Serial]Multus", func() {
 
 		Context("VirtualMachineInstance with Linux bridge plugin interface", func() {
-			table.DescribeTable("should be able to ping between two vms", func(interfaces []v1.Interface, networks []v1.Network, ifaceName string) {
+			getInterfaceAddressByNetworkName := func(vmi *v1.VirtualMachineInstance, networkName string) (string, error) {
```
Can't you just pass the vmiName as a parameter instead of the whole VMI ? AFAIU, you're not using any other parameter.

I also miss the word `IP` in the name of this function. `getIfaceIPByNetworkName` maybe ?

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -332,26 +353,55 @@ var _ = SIGDescribe("[Serial]Multus", func() {
 		Context("VirtualMachineInstance with Linux bridge plugin interface", func() {
-			table.DescribeTable("should be able to ping between two vms", func(interfaces []v1.Interface, networks []v1.Network, ifaceName string) {
+			getInterfaceAddressByNetworkName := func(vmi *v1.VirtualMachineInstance, networkName string) (string, error) {
+				vmi, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmi.Name, &metav1.GetOptions{})
```
You should bail out early on error here.

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -332,26 +353,55 @@ var _ = SIGDescribe("[Serial]Multus", func() {
+			getInterfaceAddressByNetworkName := func(vmi *v1.VirtualMachineInstance, networkName string) (string, error) {
+				vmi, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmi.Name, &metav1.GetOptions{})
+
+				ifaceAddress := ""
```
You don't need this variable; you can:

```golang
for _, iface := range vmi.Status.Interfaces {
    if iface.Name == networkName {
        return iface.IP, nil
    }
}
return "", fmt.Errorf("couldn't find iface %s on vmi %s", networkName, vmiName)
```

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -332,26 +353,55 @@ var _ = SIGDescribe("[Serial]Multus", func() {
+				ipAddr := ""
+				if staticIPVm2 == "" {
+					By("will ping secondary interface")
+					ipAddr, err = getInterfaceAddressByNetworkName(vmiTwo, networks[1].Name)
```
You depend here of a network slice w/ 2 items or more. I miss that assertion somewhere (while I do realize that it's an input of your test).

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -1485,3 +1535,32 @@ func cloudInitNetworkDataWithStaticIPsByDevice(deviceName, ipAddress string) str
 	return networkData
 }
+
+func configureAlpineInterface(vmi *v1.VirtualMachineInstance, ifaceName, staticIP string) error {
```
I also miss the word IP in the name.

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -1485,3 +1535,32 @@ func cloudInitNetworkDataWithStaticIPsByDevice(deviceName, ipAddress string) str
+	return configInterface(vmi, ifaceName, staticIP)
+}
+
+func configureAlpineInterfacesWithDynamicAddress(vmi *v1.VirtualMachineInstance, ifacesNames ...string) error {
```
`activate/enableDHCPOnVMInterfaces` maybe ? Your proposal is quite ok as is, not sure this is worth it.

maiqueb reviewed (changes requested):
Just some minor nits.

AlonaKaplan commented on pkg/virt-handler/vm.go:
```diff
@@ -874,18 +874,16 @@ func (d *VirtualMachineController) updateVMIStatus(origVMI *v1.VirtualMachineIns
-						IPs:  podIface.PodIPs,
-					}
-					interfaces = append(interfaces, ifc)
+				podIface, err := d.getPodInterfacefromFileCache(vmi, network.Name)
```
Please add a unit test.

maiqueb replied:
I'd take this a step further - we need more than one. To be on the safe side of things, I'd like to a see: 
  - a simple unit test where it shows you picking up the right IP on a secondary interface.
  - a unit test where we handle the basic scenario + have guest agent. I am expecting the guest agent to just override whatever is cached - could this be an issue ?

alonSadan reviewed (commented):
Thanks @maiqueb  and @AlonaKaplan .
[I added the two unit-tests and fixed most of what you asked me](https://github.com/kubevirt/kubevirt/compare/01525c33bcd672188d25ced737770b145a994023..cf1b2efadb120e48b38260c78d9ffab5b2547525).
I only didn't understand why should I check if there are two networks...

alonSadan commented on pkg/virt-handler/vm.go:
```diff
@@ -874,19 +874,22 @@ func (d *VirtualMachineController) updateVMIStatus(origVMI *v1.VirtualMachineIns
+					return err
+				}
+
+				if podIface != nil {
```
I also had to add this check, because sometimes, [at least fior SRIOV](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm_test.go#L2709) , we don't always cache the iface

alonSadan replied:
I am not sure, but that's why this `if` is required

alonSadan commented:
rebased on master

alonSadan commented on pkg/virt-handler/vm.go:
```diff
@@ -878,19 +878,22 @@ func (d *VirtualMachineController) updateVMIStatus(origVMI *v1.VirtualMachineIns
 			// Set Pod Interface
 			interfaces := make([]v1.VirtualMachineInstanceNetworkInterface, 0)
 			for _, network := range vmi.Spec.Networks {
-				if network.NetworkSource.Pod != nil {
```
The first part of the commit for sure this is a better wording. 
I would like also to keep the second part of the commit that explains why this change is required.

alonSadan reviewed (commented):
Thank you @maiqueb .
I addressed your comments.

maiqueb reviewed (commented):
Seems the test you've added is failing in the coverage checks: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/5826/pull-kubevirt-goveralls/1413153032547340288#1:build-log.txt%3A278

I do not understand how it passes on unit tests but fails on coverage.

Could it be related to the table ? (not sure at all).

alonSadan commented:
[rebased on master](https://github.com/kubevirt/kubevirt/compare/242035fc6e043236de7e1c6dedee0c30e24da6a2..cca88df4f10e798e284e10ddcb47c5803af0a3ce)

maiqueb commented on pkg/virt-handler/vm_test.go:
```diff
@@ -2303,57 +2303,116 @@ var _ = Describe("VirtualMachineInstance", func() {
-			vmiFeeder.Add(vmi)
-			domainFeeder.Add(domain)
+				podCacheSecondaryInterface := makePodCacheInterface(secondaryInterfaceName, secondaryInterfaceIp, secondaryInterfaceIp, secondaryInterfaceIpv6)
+				err = controller.networkCacheStoreFactory.CacheForVMI(vmi).Write(secondaryInterfaceName, podCacheSecondaryInterface)
+				Expect(err).ToNot(HaveOccurred())
```
nit: you can inline these 2.

alonSadan replied:
Done.

maiqueb commented on pkg/virt-handler/vm_test.go:
```diff
@@ -2303,57 +2303,116 @@ var _ = Describe("VirtualMachineInstance", func() {
-			controller.Execute()
-			testutils.ExpectEvent(recorder, VMIStarted)
-			Expect(controller.podInterfaceCache.Size()).To(Equal(1))
+			table.DescribeTable("should report on all networks in status", func(overrideWithGuestAgent bool) {
```
I would drop the "on".

alonSadan replied:
Done.

maiqueb commented on pkg/virt-handler/vm_test.go:
```diff
@@ -2303,57 +2303,116 @@ var _ = Describe("VirtualMachineInstance", func() {
+				testutils.ExpectEvent(recorder, VMIStarted)
+				Expect(controller.podInterfaceCache.Size()).To(Equal(2))
+			},
+				table.Entry("with cached pod interface info", false),
```
"using info from the pod interface" ?

alonSadan replied:
Done.

maiqueb commented on tests/network/vmi_multus.go:
```diff
@@ -335,26 +356,59 @@ var _ = SIGDescribe("[Serial]Multus", func() {
+
+			table.DescribeTable("should be able to ping between two vms", func(interfaces []v1.Interface, networks []v1.Network, ifaceName, staticIPVm1, staticIPVm2 string) {
+				if staticIPVm2 == "" || staticIPVm1 == "" {
+					ipam := generateIPAMConfig("host-local", "10.1.1.0/24")
+					Expect(createBridgeNetworkAttachementDefinition("linux-bridge-net-ipam", 0, ipam))
```
nit: you can inline these 2.

alonSadan replied:
I prefer not to because it will make the line to long I think.

I also forgot the `To(Succeed())` part , so I added it.

maiqueb reviewed (commented):
Nearly there, pretty much all I have are nits.

alonSadan commented on pkg/virt-handler/vm_test.go:
```diff
@@ -2303,57 +2303,116 @@ var _ = Describe("VirtualMachineInstance", func() {
+				secondaryInterfaceIpv6 = "fd10:244::8c4e"
+				domain = api.NewMinimalDomainWithUUID("testvmi", vmiTestUUID)
+
+				vmi.UID = vmiTestUUID
+				vmi.ObjectMeta.ResourceVersion = "1"
+				vmi.Spec.Networks = []v1.Network{
+					{
+						Name:          interfaceName,
+						NetworkSource: v1.NetworkSource{Pod: &v1.PodNetwork{}},
+					},
+					{
+						Name: secondaryInterfaceName,
+						NetworkSource: v1.NetworkSource{
+							Multus: &v1.MultusNetwork{
+								NetworkName: "multus",
+							},
```
> would you be kind enough to encapsulate this in a function ?

Done.
> 
> BTW: I wonder; why do you need to set the `ObjectMeta.ResourceVersion` ?

I am not sure. It was here before my change, so I didn't want to mess with it

alonSadan commented on pkg/virt-handler/vm_test.go:
```diff
@@ -2303,57 +2303,116 @@ var _ = Describe("VirtualMachineInstance", func() {
-			vmi.Status.Phase = v1.Scheduled
-			vmi.Status.Interfaces = make([]v1.VirtualMachineInstanceNetworkInterface, 0)
+				}
+				vmi.Status.Phase = v1.Scheduled
+				vmi.Status.Interfaces = make([]v1.VirtualMachineInstanceNetworkInterface, 0)
```
What is the relation between setting the phase and setting the interfaces that justifies the encapsulation  ?
or in other words, how can you generalize it?

maiqueb replied:
not sure I follow; you can add the phase as an input, and hardcode the `ResourceVersion` and the `UID` within your brand new helper func.

Does this help you ?

alonSadan replied:
It does :) 
What I meant is that defining the vmi networks, and defining it's phase and resource version are two different things, so putting it in a single helper seemed a bit weird to me. But after  rethinking I think it is fine.
Done.

maiqueb replied:
you do make a good point; you can compose your helpers in any way that makes sense - e.g. 'createSimpleVMI' + 'setDefaults' = 'createVmiWithXYZAndDefaults' .

All of the above names can be better, of course :)

alonSadan replied:
I changed the name to `getScheduledVmiWithPodNetworkAndMultusNetwork` what do you think?

alonSadan commented on pkg/virt-handler/vm_test.go:
```diff
@@ -2303,57 +2303,116 @@ var _ = Describe("VirtualMachineInstance", func() {
-			err = controller.networkCacheStoreFactory.CacheForVMI(vmi).Write(interfaceName, podCacheInterface)
-			Expect(err).ToNot(HaveOccurred())
+				podCacheInterface := makePodCacheInterface(interfaceName, interfaceIp, interfaceIp, interfaceIpv6)
+				err = controller.networkCacheStoreFactory.CacheForVMI(vmi).Write(interfaceName, podCacheInterface)
+				Expect(err).ToNot(HaveOccurred())
```
Done.

alonSadan commented on pkg/virt-handler/vm_test.go:
```diff
@@ -2303,57 +2303,116 @@ var _ = Describe("VirtualMachineInstance", func() {
+				Expect(controller.podInterfaceCache.Size()).To(Equal(2))
+			},
+				table.Entry("with cached pod interface info", false),
+				table.Entry("with guest agent interface info", true),
```
I agree. Done.

alonSadan reviewed (commented):
Thanks Miguel. Addressed your comments 
https://github.com/kubevirt/kubevirt/compare/cca88df4f10e798e284e10ddcb47c5803af0a3ce..8578c4906d79e5ed5e38f1af1db5af7b05e10595

maiqueb reviewed (approved):
Thanks.

alonSadan commented:
I noticed that in the unit test I added, I did not defined the interfaces in `domain.Spec.Devices.Interfaces`. 
In this case the interfaces are added to the status [by iterating a map](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm.go#L979) which [by definition](https://golang.org/ref/spec#Map_types) is an un-ordered data structure. I assumed in the tests that the interfaces will be ordered , and in some rare cases the order I assumed [was wrong]()https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/5826/pull-kubevirt-unit-test/1417399225620107264 what casued the tests to falke. 

Now the interfaces are defined also in `domain.Spec.Devices.Interfaces`, [and their order is set](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm.go#L962), and I hope it won't flake anymore.

alonSadan commented:
I [inserted phase and resourceversion](https://github.com/kubevirt/kubevirt/compare/e71f1f1f1117e9f3bec854357ca23a96a2dbab57..919c5e80139398382f7888fd55c496163f4d3070) to the test helper nad [changed it's name to suit](https://github.com/kubevirt/kubevirt/compare/919c5e80139398382f7888fd55c496163f4d3070..905be5a4b4bf845852a38c882ff1f327ca47d373) the changes.

AlonaKaplan commented on tests/network/vmi_multus.go:
```diff
@@ -335,26 +356,59 @@ var _ = SIGDescribe("[Serial]Multus", func() {
 				By(fmt.Sprintf("checking virtual machine interface %s state", ifaceName))
 				Expect(checkInterface(vmiTwo, ifaceName)).To(Succeed())
 
+				ipAddr, err := cidrToIP(staticIPVm2)
```
Shouldn't this be done only if `staticIPVm2 !=""`?

alonSadan replied:
Right! Thanks for the catch.
Done

AlonaKaplan commented on tests/network/vmi_multus.go:
```diff
@@ -1488,3 +1542,32 @@ func cloudInitNetworkDataWithStaticIPsByDevice(deviceName, ipAddress string) str
+}
+
+func activateDHCPOnVMInterfaces(vmi *v1.VirtualMachineInstance, ifacesNames ...string) error {
+	interfacesConfig := "auto lo\\niface lo inet loopback\\n\\n"
```
Why do we need to configure the `lo`?

alonSadan replied:
In the [alpine official docs](https://wiki.alpinelinux.org/wiki/Configure_Networking#Loopback_Configuration_.28Required.29) it states:
> Loopback Configuration (Required)
Note: The loopback configuration must appear first in /etc/network/interfaces to prevent networking issues.

So I followed that.

AlonaKaplan reviewed (commented):
Overall great! Just two tiny questions.

alonSadan commented on tests/network/vmi_multus.go:
```diff
@@ -1488,3 +1542,32 @@ func cloudInitNetworkDataWithStaticIPsByDevice(deviceName, ipAddress string) str
 	return networkData
 }
+
+func configureAlpineInterfaceIP(vmi *v1.VirtualMachineInstance, ifaceName, staticIP string) error {
```
Based on [alpine official  docs](https://wiki.alpinelinux.org/wiki/Configure_Networking#IPv4_DHCP_Configuration) I assumed this is relevant only for alpine. 
(A similar process to what's been done in `activateDHCPOnVMInterfaces`  can be done also to configure static IP, but for simplicity sake I chose to use the existing `configInterface` which also works.)
It is a bit cumbersome, but another name I can think of is `configureAlpineInterfaceWithStaticOrDynamicIP` but I am not sure it is better. What do you think?

AlonaKaplan replied:
Maybe keep `configureAlpineInterfaceIP` and just add a comment that if `staticIP` is empty the interface would set to get dynamic ip.

alonSadan replied:
[Done](https://github.com/kubevirt/kubevirt/compare/da8c102d7df19bd8aa5f3a4a812c70a377234aaf..1633d70f1d964cadfc8666f83d3f0a049bf7ae0a)

alonSadan commented:
Thanks @AlonaKaplan 
[addressed](https://github.com/kubevirt/kubevirt/compare/905be5a4b4bf845852a38c882ff1f327ca47d373..da8c102d7df19bd8aa5f3a4a812c70a377234aaf) and replied to your comments

AlonaKaplan commented:
Thanks!
/approve

alonSadan commented:
/retest
